### PR TITLE
Feature/0202 class menu

### PR DESCRIPTION
**Pull request by BenTsai85**
# Features
1. Modify all files with Redux (account manipulation, classes manipulation, navigation)
2. Integrate React-Navigation with Redux (solve Known Issues of 0200)
3. Add Account and Classes data to Store, so when Account data is updated, username displayed in Drawer is updated too (Bug/0004 fixed)
4. Move manipulation of storage data to actionCreators (async actionCreator)
5. Hide Login Page when automatically login (originally Login Page would display for a moment)
6. Display welcome message without Classes data
7. Mock data of Classes
8. Layout of Classes list
9. Add Delete button for Classes, and on pressed, update new Classes data to local storage and Store
10. Hide Delete button when scrolling
11. Load Account and Classes data from local storage during App initialization
12. Update Classes order on one Class pressed

# Test Plans
- [x] All navigation behavior works as original (including this.props.navigation.navigate('DrawerClose')).
- [x] Go to EditProfile and modify Account data. Username is updated in Drawer.
- [x] Login and EditProfile works as original.
- [x] No Login Page displayed when automatic login.
- [x] After pressing Delete button, the Class is removed for UI. When all Classes are deleted, display welcome message.
- [x] Delete button is scrolled back when scrolling Classes list or swiping open Delete button of other Classes
- [x] Login and Classes data is successfully loaded during App initialization (Automatic Login and Classes are displayed)
- [x] Update Classes order on one Class pressed
- [ ] Load Classes data from local storage instead of Mock data

# Known Issues

# Notes

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -1,32 +1,78 @@
...
+  'root',
+  state => state.nav,
+)
+const addListener = createReduxBoundAddListener('root')
+const store = createStore(reducer, applyMiddleware(middleware))
+
+const initStore = async () => {
+  // Load data from local storage
+  store.dispatch({
+    type: 'classMenu/classList/get',
```
Comment (tom6311tom6311) [inline App.js]: Action types 建議:
1. 用大寫命名 (ex. GET_CLASS_LIST)
2. 拉出成一個獨立的檔案, ex:
--actions/
----ClassList/
------ClassList.action.js
------ClassList.type.js

File: App.js [hunk 2]
```diff
@@ -1,32 +1,78 @@
...
+    payload: [{
```
Comment (tom6311tom6311) [inline App.js]: Mock data建議放到一個獨立的json檔，用fetch方式load進來

File: App.js [hunk 3]
```diff
@@ -1,32 +1,78 @@
...
+}
+
+Root.propTypes = {
+  dispatch: PropTypes.func.isRequired,
+  nav: PropTypes.object.isRequired,
+}
+
+const RootWithNavigation = connect(mapStateToProps)(Root)
+
+export default class App extends Component {
```
Comment (tom6311tom6311) [inline App.js]: 每個檔案最好只放一個Component，因此這裡建議將App和Root分開

File: src/actions/account.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import {
```
Comment (tom6311tom6311) [inline src/actions/account.js]: 建議讓檔名包含角色功能(ex. Account.action.js or AccountAction.js)

File: src/actions/classMenu.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import {
+  AsyncStorage,
+  Alert,
+} from 'react-native'
+
+
+export default {
+  classList: {
+    get: async () => {
```
Comment (tom6311tom6311) [inline src/actions/classMenu.js]: 建議不要使用'get'或'set'這種過於簡單的名稱

File: src/components/ClassItem.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  Image,
+  TouchableHighlight,
+} from 'react-native'
+import PropTypes from 'prop-types'
+import styles from './styles/ClassItem.styles'
+import DeleteImage from '../../asset/delete.png'
+
+class ClassItem extends Component {
+  componentDidMount() {
+    this.setState({ ref: this.ref })
```
Comment (tom6311tom6311) [inline src/components/ClassItem.js]: 有setState的地方都必須要做成action creator

File: src/reducers/index.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+const reducerMap = {
+  nav: nav.reducerMap,
+  classMenu: classMenu.reducerMap,
+  account: account.reducerMap,
+  initComplete: state => ({ ...state, initComplete: true }),
+}
+
+export default (state, action) => {
+  let newState = handleActions(reducerMap, initialState)(state, action)
+  if (action.type.startsWith('Navigation/')) {
```
Comment (tom6311tom6311) [inline src/reducers/index.js]: 建議用switch的方式決定action type對應要做的邏輯

File: App.js (hunk 1 above)
Comment (tom6311tom6311) [reply App.js]: ok

File: src/actions/classMenu.js (hunk 5 above)
Comment (tom6311tom6311) [reply src/actions/classMenu.js]: ok

File: src/reducers/index.js (hunk 7 above)
Comment (tom6311tom6311) [reply src/reducers/index.js]: ok

File: App.js (hunk 2 above)
Comment (awinder0230) [reply App.js]: ok

**Pull request merged**